### ISSUE TITLE
Uppercase UPN's to lowercase usernames without error

### ISSUE THREAD

**Issue by mmulrthelp** [microsoft/o365-moodle#2073]
Hi @weilai-irl 

This closed issue mentions what I am seeing. It says it was fixed in a previous release https://github.com/microsoft/o365-moodle/issues/1388

![image](https://user-images.githubusercontent.com/16646247/172832232-ccd8d7be-2aea-40f2-b028-c74626e239c5.png)

I'm not sure where this is sat, can it be fixed so if a user in Active Directory has a uppercase character in their UPN, force that to be lowercase and create the user in that format. At the moment if the sync gets to any such users it just drops out like above.

If you cannot make out the above screenshot there is a 'x' in the username and on AD this 'x' is a 'X'.

Many thanks,
Ray.

**Comment by weilai-irl** [comment]
Hi @mmulrthelp,

I tried to reproduced the issue in my test environment but didn't manage to get exact that. I'll need a bit of extra information from you.

- It looks like you have a Moodle account created for the Azure AD user already, so the user was actually "matched". The error occurred when trying to update user profile. I have tested the case when the matching Moodle user doesn't exist and let the user sync task creating the Moodle account, and the account was created successfully (with lowercase usernames) in this case.
- The first error "The username must be in lower case" is out of the scope of the plugins. Moodle requires usernames to be lowercase letters only. Even if you have the "extended characters" optional enabled, it will still only allow lowercase letters plus some special characters, and not upper case letters. So if you try to update of details of a Moodle account that has uppercase letters in the username, you will always get a "The username must be in lower case" error.
- With that in mind, please review how your Moodle accounts were created, and ensure that there are no uppercase letters in usernames.
- In my attempts, I managed to get the first error ("The username must be in lower case"), but failed to get the second error "Error writing to database". It looks like as soon as the task realised that I was tried to update a Moodle user with uppercase letters in the username, it threw the error and stopped execution at this point; but this doesn't seem to be the case in your task run. Could you turn debugging on and show the details of the SQL statement that it tried to run when getting the "Error writing to database" error please.

Regards,
Lai

**Comment by mmulrthelp** [comment]
Hi @weilai-irl the "Error writing to database" message that I was seeing is being cause by the connection to database timing out whilst the sync is running, the sync is taking 1.5hrs to do a full sync. I've asked our hosting partners to check on the timeout and see if that can be increased. This should fix this.

We've solved the issue on our end of having username with uppercase characters, so should no longer hitting this as a issue. We did this by forcing MIM (Microsoft Identity Manager) to only store lowercase character in the username / UPN. Writing this here in case it helps anyone else with the same issue.

I will close this issue - Many thanks for investigating on your end.
